packages/yarnpkg-cli/sources/main.ts

```diff
@@ -151,6 +151,11 @@ export async function main({binaryVersion, pluginConfiguration}: {binaryVersion:
 
         if (iAmHere !== iShouldBeHere) {
           process.chdir(cwd);
+
+          const cwdIndex = process.argv.findIndex(arg => arg === `--cwd`);
+          if (cwdIndex !== -1)
+            process.argv.splice(cwdIndex, 2);
```

Comment:
This wont be able to handle the following cases:
```
yarn --cwd="./foo"
yarn --cwd ./foo --cwd ./foo/bar
```

@arcanis Would it be possible for Clipanion to serialize a `Command` instance without a specific flag (`--cwd`)?

Reply:
I've read the `Clipanion` document, but I couldn't find a way to do it.

I made a fix first, and if there is a better way based on subsequent discussions I will be happy to modify it again 😀.

https://github.com/yarnpkg/berry/pull/4997/commits/c0231b615798ad2104a6fadc807ee8036f0ce023

Reply:
By the way, https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-cli/sources/main.ts#L146 always gets the last one when there are multiple `--cwd`.

```
cd ~
yarn --cwd ./foo --cwd ./bar exec pwd 
# print: ~/bar/foo
```

so I follow this from the last one to delete, but I'm not sure if this is expected (I personally rarely use `-cwd` in the command line 😢)

Reply:
> By the way, https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-cli/sources/main.ts#L146 always gets the last one when there are multiple --cwd.

Yup, that is expected in this case.

> so I follow this from the last one to delete, but I'm not sure if this is expected

We need to remove all of them except in some cases where it might be part of a [`Proxy`](https://mael.dev/clipanion/docs/options#proxies) or [`Rest`](https://mael.dev/clipanion/docs/options#rests), for example in
```sh
yarn --cwd ./foo run test --cwd="bar"
```
we can only remove the first one.
Clipanion has the information required to know which one can be removed so it needs to be done there.

Reply:
That makes sense, I learned, thanks!

Reply:
I don't imagine a nice API for that 🤔

The cause isn't clear to me from the PR description, but perhaps a nicer hack would be to keep a global variable storing whether we actually check `--cwd` or not? Once we do the `chdir` we set it to `false`, and hopefully it then won't be checked again (but I think it'd be good to explain in the PR why it's applied twice in the first place).

Reply:
here's what happened with this issue🥸

```
project (I ran the command `yarn --cwd="./foo" exec pwd` here)
|- foo
|    |- foo
|        |- foo
...
```

The program execution flow ([main.ts](https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-cli/sources/main.ts)) is as follows:

1. Execute the run function (L40).
2. Initialize Clipanion (L41).
3. Call the exec function (L48).
4. The condition iAmHere !== iShouldBeHere is satisfied (L152).
5. Call the run function (L154).
6. [Repeats execution until an error occurs.]

In this process, because the `--cwd` in `process.argv` is preserved, it keeps running repeatedly until it fails to find the directory.

For reference, this is what I observed when executing main.ts#L151:

```javascript
console.log({
  cwd,
  iAmHere,
  iShouldBeHere,
  argv: process.argv,
});
```
![image](https://github.com/yarnpkg/berry/assets/25342059/5dd96ecd-1a9f-4f9f-a584-32aaf142c5ba)

===

I have forgotten why I initially thought that global Yarn installation was necessary to reproduce this issue, However, according to the discussion later in the issue, this should be wrong🥲